elf-cleaner.cpp

```diff
@@ -328,15 +348,50 @@ int main(int argc, char **argv)
 		supported_dt_flags_1 = (DF_1_NOW | DF_1_GLOBAL | DF_1_NODELETE);
 	}
 
+	int num_threads = cpucount();
+	argmatch(argv, argc, "-jobs", "--jobs", 1, &num_threads, &skip_args);
+	if (num_threads <= 0)
+		num_threads = 1;
+
 	if (argmatch(argv, argc, "-dry-run", "--dry-run", 3, NULL, &skip_args))
 		dry_run = true;
 
 	if (argmatch(argv, argc, "-quiet", "--quiet", 3, NULL, &skip_args))
 		quiet = true;
 
-	for (int i = skip_args+1; i < argc; i++) {
-		if (parse_file(argv[i]) != 0)
-			return 1;
+	std::vector<pthread_t> thread(num_threads);
+	std::vector<struct thread_data> data(num_threads);
+	int num_files = argc - (skip_args+1);
+	/* Tasks per thread, rounded down */
+	int tasks_per_thread = (num_files / num_threads);
+
+	for (int i = 0; i<num_threads; i++) {
+		if (i < num_files % num_threads) {
+			/* Distribute 'extra' jobs (since we rounded
+			   down tasks_per_thread) */
+			data[i].argv = &argv[skip_args+1];
+			data[i].argc = tasks_per_thread + 1;
+			skip_args += tasks_per_thread + 1;
+		} else {
+			data[i].argv = &argv[skip_args+1];
+			data[i].argc = tasks_per_thread;
+			skip_args += tasks_per_thread;
+		}
+	}
+
+	/* Launch Threads */
+	for (int i = 0; i<num_threads; i++) {
+		if (data[i].argc == 0)
+			break;
+		pthread_create(&thread[i], NULL, &file_parse_loop, &data[i]);
```

Comment:
I guess this is too ugly. We probably should be relying on C++11's [`std::thread`](https://en.cppreference.com/w/cpp/thread/thread) implementation instead of creating the thread ourselves using the lower level pthread library. Also instead of going this way and dividing all the input files equally among the available CPUs, it would be much better if we:
1. First spawn a number of threads (either as specified by the user, or upon number of CPUs).
2. Have a sort of atomic object which the called thread will call `std::atomic::notify_all()` just before completing
3. The main thread will be waiting for threads to finish using `std::atomic::wait`, and spawn a new thread as soon as one finishes.

This should also increase the performance improvements in your benchmarks as not all CPUs might perform in the same fashion I guess. And also will definitely improve in conditions where the user input contains a lot of small binaries with a large binary

Reply:
I made the multithreaded implementation with std::thread.

Here's the repo: https://github.com/XniceCraft/termux-elf-cleaner/tree/multithreading

Reply:
@XniceCraft that is great! Would you like to open a pull request to replace this one?